resolwe_runtime_utils.py

```diff
@@ -170,71 +175,138 @@ def _check_response(response):
-                #
-                # TODO: fix when we support Python 3.8 and later. See
-                # dirs_exist_ok argument to copytree method.
-                _copy_file_or_dir(source.glob("*"))
-        elif source.is_file():
-            if not destination.parent.is_dir():
-                destination.parent.mkdir(parents=True)
-            # Use copy2 to preserve file metadata, such as file creation
-            # and modification times.
-            shutil.copy2(str(source), str(destination))
-
-
-def _copy_if_exists(data):
-    """Check if data output represents file.
-
-    And copy the file to the shared filesystem if this is the case.
+                raise RuntimeError(
+                    "While collecting entries: {} must be either file of directory.".format(
+                        entry
+                    )
+                )
+        return total_size
+
+    processed_dirs = set()
+    processed_files = set()
+    entry_path = Path(entry)
+    entry_size = get_entries_size([entry_path], processed_files, processed_dirs)
+    references_size = get_entries_size(
+        (Path(reference) for reference in references), processed_files, processed_dirs
+    )
+    # Upload files in chunks. If possible avoid creation of a giant
+    # list when number of referenced files is huge: its size could be
+    # over half a milion in special cases.
+    for group in grouper(processed_files, UPLOAD_FILE_BATCH_SIZE):
+        send_message(
+            command(
+                "upload_files",
+                [str(entry) for entry in group if entry is not None],
+            )
+        )
+    for group in grouper(processed_dirs, UPLOAD_FILE_BATCH_SIZE):
+        send_message(
+            command(
+                "upload_dirs",
+                [str(entry) for entry in group if entry is not None],
+            )
+        )
+    return (entry_size, references_size)
+
+
+def _determine_value_type(value):
+    """Determine the type of the output value.
+
+    This is mostly playing hide and seek with the old code. Sometimes the
+    files and directories are stored via save_file/dir methods and sometimes
+    the plain save is used with value being JSON dict representing file/dir.
+
+    In addition the name of the file could be stored in a value as a string.
+    In such case the expected behaviour is to read the content of referenced
+    file and send it instead.
     """
-    if isinstance(data, dict):
-        if "file" in data:
-            _copy_file_or_dir(chain(data.get("refs", []), (data["file"],)))
-        if "dir" in data:
-            _copy_file_or_dir(chain(data.get("refs", []), (data["dir"],)))
+    determined_type = OutputType.Value
+    if isinstance(value, dict) and any(entry in value for entry in ("file", "dir")):
+        determined_type = OutputType.FileDir
+    elif isinstance(value, str) and Path(value).is_file():
+        determined_type = OutputType.Storage
+    return determined_type
```

Comment:
You don't need `determine_value` variable, just return the value and put `OutputType.Value` at the end.